grove/measurements/estimation.py

```diff
@@ -0,0 +1,170 @@
+"""
+Utilities for estimating expected values of Pauli terms given pyquil programs
+"""
+import numpy as np
+from pyquil.paulis import (PauliSum, PauliTerm, commuting_sets, sI,
+                           term_with_coeff)
+from pyquil.quil import Program
+from pyquil.gates import RY, RX
+
+
+class CommutationError(Exception):
+    pass
+
+
+def remove_imaginary_terms(pauli_sums):
+    """
+    Remove the imaginary component of each term in a Pauli sum
+
+    :param PauliSum pauli_sums: The Pauli sum to process.
+    :return: a purely hermitian Pauli sum.
+    :rtype: PauliSum
+    """
+    if not isinstance(pauli_sums, PauliSum):
+        raise TypeError("not a pauli sum. please give me one")
+    new_term = sI(0) * 0.0
+    for term in pauli_sums:
+        new_term += term_with_coeff(term, term.coefficient.real)
+
+    return new_term
+
+
+def get_rotation_program(pauli_term):
+    """
+    Generate a rotation program so that the pauli term is diagonal
+
+    :param PauliTerm pauli_term: The Pauli term used to generate diagonalizing
+                                 one-qubit rotations.
+    :return: The rotation program.
+    :rtype: Program
+    """
+    meas_basis_change = Program()
+    for index, gate in pauli_term:
+        if gate == 'X':
+            meas_basis_change.inst(RY(-np.pi / 2, index))
+        elif gate == 'Y':
+            meas_basis_change.inst(RX(np.pi / 2, index))
+        elif gate == 'Z':
+            pass
+        else:
+            raise ValueError()
+
+    return meas_basis_change
+
+
+def get_parity(pauli_terms, bitstring_results):
+    """
+    Calculate the eigenvalues of Pauli operators given results of projective measurements
+
+    The single-qubit projective measurement results (elements of
+    `bitstring_results`) are listed in physical-qubit-label numerical order.
+
+    An example:
+
+    Consider a Pauli term Z1 Z5 Z6 I7 and a collection of single-qubit
+    measurement results corresponding to measurements in the z-basis on qubits
+    {1, 5, 6, 7}. Each element of bitstring_results is an element of
+    :math:`\{0, 1\}^{\otimes 4}`.  If [0, 0, 1, 0] and [1, 0, 1, 1]
+    are the two projective measurement results in `bitstring_results` then
+    this method returns a 1 x 2 numpy array with values [[-1, 1]]
+
+    :param List pauli_terms: A list of Pauli terms operators to use
+    :param bitstring_results:
+    :return:
```

Comment:
Please still improve the documentation  for `bitstring_results` and the `:return:` statement.